catalyst-gateway/bin/src/service/api/cardano/cip36/endpoint.rs

```diff
@@ -1,29 +1,79 @@
 //! Implementation of the GET `/cardano/cip36` endpoint
 
-use std::time::Duration;
-
-use poem::http::HeaderMap;
-use tokio::time::sleep;
+use poem::http::{HeaderMap, StatusCode};
+use tracing::error;
 
+use self::cardano::{hash28::HexEncodedHash28, query::stake_or_voter::StakeAddressOrPublicKey};
 use super::{
     cardano::{self},
-    response, NoneOrRBAC, SlotNo,
+    filter::{get_registration_given_stake_key_hash, get_registration_given_vote_key, snapshot},
+    response, SlotNo,
+};
+use crate::{
+    db::index::session::CassandraSession,
+    service::{
+        api::cardano::cip36::response::AllRegistration,
+        common::{self},
+    },
 };
-use crate::service::common::{self};
 
 /// Process the endpoint operation
 pub(crate) async fn cip36_registrations(
-    _lookup: Option<cardano::query::stake_or_voter::StakeOrVoter>, _asat: Option<SlotNo>,
+    lookup: Option<cardano::query::stake_or_voter::StakeOrVoter>, asat: Option<SlotNo>,
     _page: common::types::generic::query::pagination::Page,
-    _limit: common::types::generic::query::pagination::Limit, _auth: NoneOrRBAC,
-    _headers: &HeaderMap,
+    _limit: common::types::generic::query::pagination::Limit, _headers: &HeaderMap,
 ) -> response::AllRegistration {
-    // Dummy sleep, remove it
-    sleep(Duration::from_millis(1)).await;
+    let Some(session) = CassandraSession::get(true) else {
+        error!("Failed to acquire db session");
+        return AllRegistration::unprocessable_content(vec![poem::Error::from_string(
+            "Connection issue",
+            StatusCode::UNPROCESSABLE_ENTITY,
+        )]);
+    };
+
+    if let Some(stake_or_voter) = lookup {
+        match StakeAddressOrPublicKey::from(stake_or_voter) {
+            StakeAddressOrPublicKey::Address(cip19_stake_address) => {
+                // Typically, a stake address will start with 'stake1',
+                // We need to convert this to a stake hash as per our data model to then find the,
+                // Full Stake Public Key (32 byte Ed25519 Public key, not hashed).
+                // We then get the latest registration or from a specific time as optionally
+                // specified in the query parameter. This can be represented as either
+                // the blockchains slot number or a unix timestamp.
+                let stake_hash: HexEncodedHash28 = match cip19_stake_address.try_into() {
+                    Ok(stake_hash) => stake_hash,
+                    Err(err) => {
+                        return response::AllRegistration::unprocessable_content(vec![
```

Comment:
This should be a `BadRequest` response.

Reply:
unprocessable content is basically the same, see [here ](https://github.com/input-output-hk/catalyst-voices/blob/main/catalyst-gateway/bin/src/service/common/responses/code_422_unprocessable_content.rs#L9)

However, I have removed it in most places that didn't make sense and replaced with internal err and handle err